### PR TITLE
errors: fully classify all errors

### PR DESCRIPTION
**Pull request by Raynos**
The `classify()` method was not exhaustive. It now is and
there are tests to ensure that it will always be exhaustive

This will be used for improving retries in hyperbahn/handler
and for improving error logging in relay_handler.

r: @kriskowal @rf

### CODE REVIEW

**Comment by kriskowal** [comment]
:ok: When @jcorbin gets back, let’s talk about disseminating this information out to the error constructors.

**Comment by rf** [comment]
cool and neat

File: node/errors.js [hunk 1]
```diff
@@ -445,21 +444,46 @@ module.exports.classify = function classify(err) {
...
+        case 'tchannel-json-handler.stringify-error.body-failed':
+        case 'tchannel-json-handler.parse-error.head-failed':
+        case 'tchannel-json-handler.stringify-error.head-failed':
+        case 'tchannel-thrift-handler.stringify-error.body-failed':
+        case 'tchannel-thrift-handler.stringify-error.head-failed':
+        case 'tchannel.request-frame-state':
+        case 'tchannel.request-already-done':
+        case 'tchannel.response-already-done':
+        case 'tchannel.response-already-started':
+        case 'tchannel.response-frame-state':
```
Comment (rf) [inline node/errors.js]: response already started etc should probably be unexpected errors

File: node/errors.js [hunk 2]
```diff
@@ -445,21 +444,46 @@ module.exports.classify = function classify(err) {
...
         case 'tchannel.connection.timeout':
         case 'tchannel.connection-stale.timeout':
             return 'Timeout';
 
-        case 'tchannel-handler.parse-error.body-failed':
-        case 'tchannel-handler.parse-error.head-failed':
+        case 'tchannel.init.call-request-before-init-request':
+        case 'tchannel.init.call-request-cont-before-init-request':
+        case 'tchannel.init.call-response-before-init-response':
+        case 'tchannel.init.call-response-cont-before-init-response':
```
Comment (rf) [inline node/errors.js]: These should probably be protocol errors

File: node/errors.js [hunk 3]
```diff
@@ -445,21 +444,46 @@ module.exports.classify = function classify(err) {
...
         case 'tchannel.duplicate-header-key':
             return 'BadRequest';
 
+        case 'tchannel.init.duplicate-init-request':
+        case 'tchannel.init.duplicate-init-response':
+        case 'tchannel.init.send-call-request-before-indentified':
+        case 'tchannel.init.send-call-request-cont-before-indentified':
+        case 'tchannel.init.send-call-response-before-indentified':
+        case 'tchannel.init.send-call-response-cont-before-indentified':
+        case 'tchannel.arg1-over-length-limit':
```
Comment (rf) [inline node/errors.js]: Over length limit sounds like a bad request

**Comment by Raynos** [comment]
Updated with some more fixes. @rf @kriskowal

File: node/errors.js [hunk 4]
```diff
@@ -445,34 +444,71 @@ module.exports.classify = function classify(err) {
         case 'tchannel.max-pending-for-service':
             return 'Declined';
 
-        case 'tchannel.timeout':
+        case 'tchannel.socket-local-closed':
+        case 'tchannel.local.reset':
+            return 'Cancelled';
```
Comment (kriskowal) [inline node/errors.js]: Whether local close or reset corresponds to a deliberate cancellation is debatable.

I do suspect that these cases should be candidates for retry.

These cases do not indicate remote health problems, so like cancellation and bad request, they should not trip circuit breaking. Perhaps we need another error class.

Whether deliberate and accidental cancellation should have the same retry semantics is the question.

File: node/errors.js [hunk 5]
```diff
@@ -445,34 +444,71 @@ module.exports.classify = function classify(err) {
...
-        case 'tchannel-handler.parse-error.body-failed':
-        case 'tchannel-handler.parse-error.head-failed':
+        case 'tchannel-handler.json.invalid-body':
+        case 'tchannel-json-handler.parse-error.body-failed':
+        case 'tchannel-json-handler.parse-error.head-failed':
+        case 'tchannel.request-frame-state':
+        case 'tchannel.request-already-done':
+        case 'tchannel-thrift-handler.parse-error.body-failed':
+        case 'tchannel-thrift-handler.parse-error.head-failed':
         case 'tchannel.checksum':
```
Comment (kriskowal) [inline node/errors.js]: Will have to talk to @jcorbin about whether we want to retry if a bit gets flipped in transit. Bad checksum is usually a bad request.

**Comment by kriskowal** [comment]
:ok:

File: node/errors.js (hunk 4 above)
Comment (jcorbin) [inline node/errors.js]: Same feedback essentially as Kris: this isn't really a cancellation user action, in production I'd expect this to happen most frequently during upgrades / restarts / etc.

File: node/errors.js [hunk 6]
```diff
@@ -445,34 +444,71 @@ module.exports.classify = function classify(err) {
...
         case 'tchannel.invalid-frame-type':
         case 'tchannel.missing-init-header':
-        case 'tchannel.null-key':
         case 'tchannel.protocol.read-failed':
         case 'tchannel.protocol.write-failed':
         case 'tchannel.unhandled-frame-type':
             return 'ProtocolError';
 
+        case 'tchannel.connection.close':
+        case 'tchannel.connection.reset':
```
Comment (jcorbin) [inline node/errors.js]: the mislabeled "cancel" socket closes above should be right here.

File: node/errors.js (hunk 4 above)
Comment (Raynos) [inline node/errors.js]: The only time these happen is when you do a graceful shutdown of a connection.

This generally only happens when you call channel.close() and that's only really done by our tests; not in production.

I agree that cancellation is not ideal. But manual closing has nothing to do with network error. As far as we know the network is healthy. Maybe it's better marked as declined due to graceful close.

Comment (jcorbin) [inline node/errors.js]: I don't think that word means what you think it means, this is no a "graceful" shutdown.  A graceful shutdown would wait for all pending requests to be finished before closing / shutting down, while at the same time not accepting any more incoming requests.

Comment (Raynos) [inline node/errors.js]: I buy that it should be a network error.

**Pull request merged**